azure-iot-device/tests/iothub/aio/test_async_clients.py

```diff
@@ -52,69 +73,126 @@ class ClientSharedTests(object):
     ):
         client = self.client_class.from_authentication_provider(auth_provider, protocol)
         assert isinstance(client, self.client_class)
         assert isinstance(client._pipeline, expected_pipeline)
 
+    @pytest.mark.it("Raises ValueError on invalid protocol input")
     @pytest.mark.parametrize("auth_provider", ["SymmetricKey"], ids=[""], indirect=True)
-    async def test_from_authentication_provider_bad_input_raises_error_transport_name(
-        self, auth_provider
-    ):
+    # TODO: handle the auth_provider fixture differently so it doesn't result in empty [] in output
+    async def test_bad_protocol_input_raises_error(self, auth_provider):
         with pytest.raises(ValueError):
             self.client_class.from_authentication_provider(auth_provider, "bad input")
 
-    async def test_instantiation_sets_on_connected_handler_in_pipeline(self, client):
-        assert client._pipeline.on_transport_connected is not None
-        assert client._pipeline.on_transport_connected == client._on_state_change
 
-    async def test_instantiation_sets_on_disconnected_handler_in_pipeline(self, client):
-        assert client._pipeline.on_transport_disconnected is not None
-        assert client._pipeline.on_transport_disconnected == client._on_state_change
+class SharedClientConnectTests(object):
+    @pytest.mark.it("Begins a 'connect' pipeline operation")
+    async def test_calls_pipeline_connect(self, client, pipeline):
+        await client.connect()
+        assert pipeline.connect.call_count == 1
 
-    async def test_instantiation_sets_on_method_request_received_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_method_request_received is not None
-        assert (
-            client._pipeline.on_transport_method_request_received
-            == client._inbox_manager.route_method_request
-        )
+    @pytest.mark.it("Waits for the completion of the 'connect' pipeline operation before returning")
+    async def test_waits_for_pipeline_op_completion(self, mocker, client, pipeline):
+        cb_mock = mocker.patch.object(async_adapter, "AwaitableCallback").return_value
+        cb_mock.completion.return_value = await create_completed_future(None)
+
+        await client.connect()
+
+        # Assert callback is sent to pipeline
+        assert pipeline.connect.call_args[1]["callback"] is cb_mock
+        # Assert callback completion is waited upon
+        assert cb_mock.completion.call_count == 1
+
+
+class SharedClientDisconnectTests(object):
+    @pytest.mark.it("Begins a 'disconnect' pipeline operation")
+    async def test_calls_pipeline_disconnect(self, client, pipeline):
+        await client.disconnect()
+        assert pipeline.disconnect.call_count == 1
+
+    @pytest.mark.it(
+        "Waits for the completion of the 'disconnect' pipeline operation before returning"
+    )
+    async def test_waits_for_pipeline_op_completion(self, mocker, client, pipeline):
```

Comment:
>test_waits_for_pipeline_op_completion [](start = 14, length = 37)

This pattern is less convoluted than the sync version, and it looks like it asserts everything correctly, but it's still a very roundabout way to do this.  You could do a similar pattern as the sync code and avoid the patching.

```
future = client.disconnect()
assert not future.done()
manually_trigger_callback()
assert future.done()
```

Reply:
No, that won't work. As mentioned in the other comment on the sync implementation, this doesn't prove that the completion happens as a direct result of the callback as opposed to just timing issues.

---
In reply to: [291329756](https://github.com/Azure/azure-iot-sdk-python-preview/pull/101#discussion_r291329756) [](ancestors = 291329756)